Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.cpp

```diff
@@ -116,63 +120,107 @@ void AcceleratedSurfaceDMABuf::RenderTargetColorBuffer::willRenderFrame() const
+    }
+
     auto* device = WebCore::GBMDevice::singleton().device();
     if (!device) {
         WTFLogAlways("Failed to create GBM buffer of size %dx%d: no GBM device found", size.width(), size.height());
         return nullptr;
     }
 
-    auto* bo = gbm_bo_create(device, size.width(), size.height(), uint32_t(WebCore::DMABufFormat::FourCC::ARGB8888), GBM_BO_USE_RENDERING);
+    struct gbm_bo* bo = nullptr;
+    uint64_t modifier = DRM_FORMAT_MOD_INVALID;
+    uint32_t flags = dmabufFormat.usage == DMABufRendererBufferFormat::Usage::Scanout ? GBM_BO_USE_SCANOUT : GBM_BO_USE_RENDERING;
+    if (dmabufFormat.modifiers[0] != DRM_FORMAT_MOD_INVALID) {
+#if HAVE(GBM_BO_CREATE_WITH_MODIFIERS2)
+        bo = gbm_bo_create_with_modifiers2(device, size.width(), size.height(), dmabufFormat.fourcc, dmabufFormat.modifiers.data(), dmabufFormat.modifiers.size(), flags);
+#else
+        bo = gbm_bo_create_with_modifiers(device, size.width(), size.height(), dmabufFormat.fourcc, dmabufFormat.modifiers.data(), dmabufFormat.modifiers.size());
+#endif
+        if (bo)
+            modifier = gbm_bo_get_modifier(bo);
+    }
+
+    if (!bo) {
+        if (dmabufFormat.usage == DMABufRendererBufferFormat::Usage::Mapping)
+            flags |= GBM_BO_USE_LINEAR;
+        bo = gbm_bo_create(device, size.width(), size.height(), dmabufFormat.fourcc, flags);
+    }
+
     if (!bo) {
         WTFLogAlways("Failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
         return nullptr;
     }
 
-    UnixFileDescriptor fd { gbm_bo_get_fd(bo), UnixFileDescriptor::Adopt };
+    Vector<UnixFileDescriptor> fds;
+    Vector<uint32_t> offsets;
+    Vector<uint32_t> strides;
     uint32_t format = gbm_bo_get_format(bo);
-    uint32_t offset = gbm_bo_get_offset(bo, 0);
-    uint32_t stride = gbm_bo_get_stride(bo);
-    uint64_t modifier = gbm_bo_get_modifier(bo);
+    int planeCount = gbm_bo_get_plane_count(bo);
 
     Vector<EGLAttrib> attributes = {
         EGL_WIDTH, static_cast<EGLAttrib>(gbm_bo_get_width(bo)),
         EGL_HEIGHT, static_cast<EGLAttrib>(gbm_bo_get_height(bo)),
         EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLAttrib>(format),
-        EGL_DMA_BUF_PLANE0_FD_EXT, fd.value(),
-        EGL_DMA_BUF_PLANE0_OFFSET_EXT, static_cast<EGLAttrib>(offset),
-        EGL_DMA_BUF_PLANE0_PITCH_EXT, static_cast<EGLAttrib>(stride),
     };
-    auto& display = WebCore::PlatformDisplay::sharedDisplayForCompositing();
-    if (modifier != uint64_t(WebCore::DMABufFormat::Modifier::Invalid) && display.eglExtensions().EXT_image_dma_buf_import_modifiers) {
-        std::array<EGLAttrib, 4> modifierAttributes {
-            EGL_DMA_BUF_PLANE0_MODIFIER_HI_EXT, static_cast<EGLAttrib>(modifier >> 32),
-            EGL_DMA_BUF_PLANE0_MODIFIER_LO_EXT, static_cast<EGLAttrib>(modifier & 0xffffffff),
-        };
-        attributes.append(std::span<const EGLAttrib> { modifierAttributes });
+
+#define ADD_PLANE_ATTRIBUTES(planeIndex) { \
```

Comment:
Is this define duplicated? Could we use some shared definition?

Reply:
No, it's the same name, but it has a different implementation. One is used in the code to export the DMA-BUF and the other in the code to import it. In both cases the macro is only used locally in that function and unset right after being used.